### PR TITLE
Fill in content

### PR DESCRIPTION
**Pull request by maroux**

### CODE REVIEW

**Comment by maroux** [comment]
https://maroux.github.io/taskhawk/ for a demo of how it'll look once deployed.

File: README.md [hunk 1]
```diff
@@ -1,4 +1,154 @@
 # Taskhawk
 
-TaskHawk is a replacement for celery that works on AWS SQS/SNS, while keeping things pretty simple and
-straightforward.
+TaskHawk is a simple asynchronous task queue that works on AWS SQS. It's reasonable to think of
```
Comment (mngo87) [inline README.md]: should mention that this can be extended to other queueing platforms, but for now it's only SQS

Comment (mngo87) [reply README.md]: Also get rid of contractions, so it is more formal

File: README.md [hunk 2]
```diff
@@ -1,4 +1,154 @@
...
+this as a replacement for Celery (Python) for people who don't need the complexity of celery, or
+want to deal with failures in Celery and/or RabbitMQ.
```
Comment (mngo87) [inline README.md]: i would remove the rabbitmq reference

File: README.md [hunk 3]
```diff
@@ -1,4 +1,154 @@
...
+
+## Why?
+
+Like a lot of other Python developers, when we needed asynchronous worker pattern for our project,
+we decided to go with Celery. Since Celery didn't have reasonable support any queue systems other
+than RabbitMQ, that was the our choice of backend. It worked well to start with, however as we
+deployed it into production and scaled the number of messages flowing through the
+system, we started to see a bunch of odd failures, such as, [worker gets stuck](https://github.com/celery/celery/issues?q=label%3A%22Worker+Hangs),
```
Comment (mngo87) [inline README.md]: worker gets stuck link does not show any issues

File: README.md [hunk 4]
```diff
@@ -1,4 +1,154 @@
...
+[broken logging](https://github.com/celery/celery/issues/4326), [RabbitMQ issues](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Component%3A+RabbitMQ+Broker%22).
+
+With Celery v4, came the promise of AWS SQS, which we love and use in other places, so we attempted
+a switch from RabbitMQ to SQS but found that the support was mostly terrible, including [exponentially
+growing duplicate messages!](https://github.com/celery/celery/issues/4002#issuecomment-327255816).
```
Comment (mngo87) [inline README.md]: +1

File: README.md [hunk 5]
```diff
@@ -1,4 +1,154 @@
...
+
+Having used celery in a bunch of projects, we realized we didn't use or need a majority of functionality
+offered by celery, such as a results backend, task monitoring, chords, groups, etc. All we needed
+was a simple and reliable queue that would invoke worker code asynchronously.
+
+Automatic relies heavily of AWS, including Simple Queue Service. While Amazon doesn't provide any
+SLA guarantees, in practice SQS scales infinitely with minimal data loss. SQS also provides at least
+once delivery guarantee. We prefer using AWS infrastructure over self-managed infrastructure wherever
+the offering matches our needs. We call this "exporting our nines (9s) to Amazon".
```
Comment (mngo87) [inline README.md]: heh

File: README.md [hunk 6]
```diff
@@ -1,4 +1,154 @@
...
+async task queue that works on AWS SQS. Since SQS gives us reliability and durability, we
+don't have to worry about the queue itself. Also having excellent API support and libraries like
+boto, we don't have to worry about the SQS interface.
+
+## What?
+
+Taskhawk is a simple library to work with SNS/SQS to dispatch and consume asynchronous tasks. It supports
+apps running on AWS Lambda as well as traditional apps. Your code needn't run in AWS, it just needs
+to be able to talk to AWS APIs. For Lambda apps, Taskhawk uses SNS to dispatch tasks which can execute
+your lambda function. All messages need to be ack-ed after successful processing. Tasks are held in
```
Comment (mngo87) [inline README.md]: can be confusing, cause you don't explicitly ack a lambda sns event

File: README.md [hunk 7]
```diff
@@ -1,4 +1,154 @@
...
+SQS queue (or internal Lambda queue) until they’re successfully executed, or until they fail a configurable
+number of times. Failed tasks are moved to a Dead Letter Queue, where they’re held for 14 days,
```
Comment (mngo87) [inline README.md]: in dlq is a recommendation (and limit), not a default of a general queue policy

File: README.md [hunk 8]
```diff
@@ -1,4 +1,154 @@
...
+and may be examined for further debugging.
+
+Taskhawk provides 4 priority queues which may be used for independent monitoring and scaling needs. In
```
Comment (mngo87) [inline README.md]: provide a brief description of priorities.  also i would make it clear this is not a `priority queue` data structure, but separate queues tagged with a priority

File: README.md [hunk 9]
```diff
@@ -1,4 +1,154 @@
...
+message id must be generated when message is initialized, and must be preserved during retries if
+attempted. All key-value pairs in `headers` must be published as SQS/SNS message attributes.
+
+Optionally, libraries may provide hooks for modifying message contents/headers before publishing,
+or for inspecting message contents/headers after consuming before processing the message. This is useful,
+for example, to inject a request id for cross-application tracing or general debugging.
+
+#### SQS
+
+Consumer should fetch SQS/SNS message attributes on de-queue but NOT use it to modify the message payload.
```
Comment (mngo87) [inline README.md]: drop SNS, more of a general message attribute

File: README.md (hunk 1 above)
Comment (maroux) [reply README.md]: adding a section for roadmap will include support for other queuing platforms there - works?

File: README.md (hunk 3 above)
Comment (maroux) [reply README.md]: good catch, bad HTML param editing skills 😛

Comment (maroux) [reply README.md]: speaking of - any other issues to add to this list?

File: README.md (hunk 5 above)
Comment (maroux) [reply README.md]: 😎

File: README.md (hunk 6 above)
Comment (maroux) [reply README.md]: good call.

File: README.md (hunk 7 above)
Comment (maroux) [reply README.md]: I mean it is default of how Taskhawk is supposed to be used - and the default provisioning setup if you use our TF

File: README.md (hunk 8 above)
Comment (maroux) [reply README.md]: Good point.

File: README.md (hunk 1 above)
Comment (mngo87) [reply README.md]: yup

**Pull request merged**